### ISSUE TITLE
Timeout in DatabricksSubmitRunOperator causes callbacks to be invoked twice

### ISSUE THREAD

**Issue by Oduig** [apache/airflow#29461]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

The `on_failure_callback` is invoked twice when a DAG fails due to a timeout. This leads to duplicate failure alerts in our Slack channels. It happens reliably and the invocations are roughly 5 seconds apart. Context and full stack trace in both invocations are identical.

### What you think should happen instead

The `on_failure_callback` should only be invoked once

### How to reproduce

Here is the DAG which triggers the effect

```
import pendulum
from datetime import timedelta
from airflow import DAG
from airflow.providers.databricks.operators.databricks import DatabricksSubmitRunOperator
import logging
import requests

dag_settings = {
    "dag_id": f"INT_minimal_example",
    "max_active_runs": 1,
    "dagrun_timeout": timedelta(minutes=1),
    "start_date": pendulum.today("UTC"),
    "default_args": {
        "owner": "airflow",
        "catchup": False
    },
    "tags": ["env:testing"],
    "on_failure_callback": (
        lambda context: [
            logging.info(f"Minimal example - failure callback with context: {context}"),
            # Since logging.info does not work for us, we add a request to RequestBin
            # See https://github.com/apache/airflow/issues/29442
            requests.post(
                "https://<removed-for-publication>.m.pipedream.net",
                json={"payload": f"Minimal example - failure callback with context: {context}"}
            )
        ]
    )
}
dag = DAG(**dag_settings)
logging.info(f"Minimal example - Created DAG {dag.dag_id}.")

DatabricksSubmitRunOperator(
    task_id="example_task",
    dag=dag,
    databricks_conn_id="databricks_default",
    spark_submit_task={
        "parameters":  [
            "--class", "com.example.Launcher",
            f"dbfs:/libraries/scala/example-fat-jar.jar"
        ]
    },
    new_cluster={
        "spark_version":
        "10.4.x-cpu-ml-scala2.12",
        "spark_env_vars": {
            "JNAME": "zulu11-ca-amd64"  # Use JDK 11
        },
        "spark_conf": {
            "spark.sql.session.timeZone": "UTC"
        },
        "aws_attributes": {
            "instance_profile_arn": "arn:aws:iam::765<removed-for-publication>:instance-profile/DatabricksExecution"
        },
        "instance_pool_id": "1011-<removed-for-publication>",
        "driver_instance_pool_id": "1011-<removed-for-publication>",
        "num_workers": 1
    }
)
```

### Operating System

composer-2.0.32

### Versions of Apache Airflow Providers

Here is the full `requirements.txt`

```
apache-airflow-providers-databricks==4.0.0
databricks-sql-connector==2.1.0
apache-beam~=2.43.0
sqlalchemy-bigquery==1.5.0
requests~=2.28.1
apache-airflow-providers-tableau==4.0.0
apache-airflow-providers-sendgrid==3.1.0
python-dotenv==0.21.0
urllib3~=1.26.8
tableauserverclient==0.23
apache-airflow-providers-http==4.1.0
# time library in airflow
pendulum==2.1.2
```

### Deployment

Composer

### Deployment details

We are running a Cloud Compose environment with image `composer-2.0.32-airflow-2.3.4`

### Anything else

Since logging inside the callback is not working for us (See #29442), here is a screenshot from RequestBin. Both invocations share the same log line, the context and stack trace is also the same.

![image](https://user-images.githubusercontent.com/3661031/218069601-dada3f0e-4494-4cf3-a274-f5d3b8270ac8.png)

Here is the complete stack trace for either invocation.

```
File "/opt/python3.8/bin/airflow", line 8, in <module>
    sys.exit(main())
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/__main__.py", line 38, in main
    args.func(args)
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/cli/cli_parser.py", line 51, in command
    return func(*args, **kwargs)
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/utils/cli.py", line 101, in wrapper
    return f(*args, **kwargs)
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/cli/commands/scheduler_command.py", line 76, in scheduler
    _run_scheduler_job(args=args)
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/cli/commands/scheduler_command.py", line 46, in _run_scheduler_job
    job.run()
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/jobs/base_job.py", line 244, in run
    self._execute()
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/jobs/scheduler_job.py", line 748, in _execute
    self.processor_agent.start()
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/dag_processing/manager.py", line 158, in start
    process.start()
  File "/opt/python3.8/lib/python3.8/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/opt/python3.8/lib/python3.8/multiprocessing/context.py", line 277, in _Popen
    return Popen(process_obj)
  File "/opt/python3.8/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/opt/python3.8/lib/python3.8/multiprocessing/popen_fork.py", line 75, in _launch
    code = process_obj._bootstrap(parent_sentinel=child_r)
  File "/opt/python3.8/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/opt/python3.8/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/dag_processing/manager.py", line 255, in _run_processor_manager
    processor_manager.start()
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/dag_processing/manager.py", line 486, in start
    return self._run_parsing_loop()
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/dag_processing/manager.py", line 600, in _run_parsing_loop
    self.start_new_processes()
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/dag_processing/manager.py", line 1003, in start_new_processes
    processor.start()
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/dag_processing/processor.py", line 194, in start
    process.start()
  File "/opt/python3.8/lib/python3.8/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/opt/python3.8/lib/python3.8/multiprocessing/context.py", line 277, in _Popen
    return Popen(process_obj)
  File "/opt/python3.8/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/opt/python3.8/lib/python3.8/multiprocessing/popen_fork.py", line 75, in _launch
    code = process_obj._bootstrap(parent_sentinel=child_r)
  File "/opt/python3.8/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/opt/python3.8/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/dag_processing/processor.py", line 155, in _run_file_processor
    result: Tuple[int, int] = dag_file_processor.process_file(
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/utils/session.py", line 71, in wrapper
    return func(*args, session=session, **kwargs)
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/dag_processing/processor.py", line 656, in process_file
    self.execute_callbacks(dagbag, callback_requests)
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/utils/session.py", line 71, in wrapper
    return func(*args, session=session, **kwargs)
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/dag_processing/processor.py", line 581, in execute_callbacks
    self._execute_dag_callbacks(dagbag, request, session)
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/utils/session.py", line 68, in wrapper
    return func(*args, **kwargs)
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/dag_processing/processor.py", line 595, in _execute_dag_callbacks
    dag.handle_callback(
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/utils/session.py", line 68, in wrapper
    return func(*args, **kwargs)
  File "/opt/python3.8/lib/python3.8/site-packages/airflow/models/dag.py", line 1178, in handle_callback
    callback(context)
  File "/home/airflow/gcs/dags/integration_test/example.py"
```

Here is the log for the Databricks task that was marked skipped, because the DAG as a whole timed out.

```
*** Reading remote log from gs://europe-west3-airflow-<removed-for-publication>-bucket/logs/dag_id=INT_minimal_example/run_id=manual__2023-02-10T09:34:38+00:00/task_id=example_task/attempt=1.log.
[2023-02-10, 09:34:42 UTC] {taskinstance.py:1172} INFO - Dependencies all met for <TaskInstance: INT_minimal_example.example_task manual__2023-02-10T09:34:38+00:00 [queued]>
[2023-02-10, 09:34:42 UTC] {taskinstance.py:1172} INFO - Dependencies all met for <TaskInstance: INT_minimal_example.example_task manual__2023-02-10T09:34:38+00:00 [queued]>
[2023-02-10, 09:34:42 UTC] {taskinstance.py:1369} INFO - 
--------------------------------------------------------------------------------
[2023-02-10, 09:34:42 UTC] {taskinstance.py:1370} INFO - Starting attempt 1 of 3
[2023-02-10, 09:34:42 UTC] {taskinstance.py:1371} INFO - 
--------------------------------------------------------------------------------
[2023-02-10, 09:34:42 UTC] {taskinstance.py:1390} INFO - Executing <Task(DatabricksSubmitRunOperator): example_task> on 2023-02-10 09:34:38+00:00
[2023-02-10, 09:34:42 UTC] {standard_task_runner.py:52} INFO - Started process 69373 to run task
[2023-02-10, 09:34:42 UTC] {standard_task_runner.py:79} INFO - Running: ['airflow', 'tasks', 'run', 'INT_minimal_example', 'example_task', 'manual__2023-02-10T09:34:38+00:00', '--job-id', '197269', '--raw', '--subdir', 'DAGS_FOLDER/integration_test/minimal_example.py', '--cfg-path', '/tmp/tmp4ffeji6u', '--error-file', '/tmp/tmpd5pieigz']
[2023-02-10, 09:34:42 UTC] {standard_task_runner.py:80} INFO - Job 197269: Subtask example_task
[2023-02-10, 09:34:43 UTC] {task_command.py:375} INFO - Running <TaskInstance: INT_minimal_example.example_task manual__2023-02-10T09:34:38+00:00 [running]> on host airflow-worker-lhbx8
[2023-02-10, 09:34:43 UTC] {taskinstance.py:1583} INFO - Exporting the following env vars:
AIRFLOW_CTX_DAG_OWNER=airflow
AIRFLOW_CTX_DAG_ID=INT_minimal_example
AIRFLOW_CTX_TASK_ID=example_task
AIRFLOW_CTX_EXECUTION_DATE=2023-02-10T09:34:38+00:00
AIRFLOW_CTX_TRY_NUMBER=1
AIRFLOW_CTX_DAG_RUN_ID=manual__2023-02-10T09:34:38+00:00
[2023-02-10, 09:34:43 UTC] {base.py:68} INFO - Using connection ID 'databricks_default' for task execution.
[2023-02-10, 09:34:43 UTC] {databricks_base.py:333} INFO - Using basic auth.
[2023-02-10, 09:34:43 UTC] {databricks.py:75} INFO - Run submitted with run_id: 102832491
[2023-02-10, 09:34:43 UTC] {databricks_base.py:333} INFO - Using basic auth.
[2023-02-10, 09:34:43 UTC] {databricks_base.py:333} INFO - Using basic auth.
[2023-02-10, 09:34:43 UTC] {databricks.py:97} INFO - example_task in run state: {'life_cycle_state': 'PENDING', 'result_state': '', 'state_message': ''}
[2023-02-10, 09:34:43 UTC] {databricks.py:98} INFO - View run status, Spark UI, and logs at <removed-for-publication>
[2023-02-10, 09:34:43 UTC] {databricks.py:99} INFO - Sleeping for 30 seconds.
[2023-02-10, 09:35:13 UTC] {databricks_base.py:333} INFO - Using basic auth.
[2023-02-10, 09:35:13 UTC] {databricks.py:97} INFO - example_task in run state: {'life_cycle_state': 'PENDING', 'result_state': '', 'state_message': 'Waiting for cluster'}
[2023-02-10, 09:35:13 UTC] {databricks.py:98} INFO - View run status, Spark UI, and logs at <removed-for-publication>
[2023-02-10, 09:35:13 UTC] {databricks.py:99} INFO - Sleeping for 30 seconds.
[2023-02-10, 09:35:43 UTC] {databricks_base.py:333} INFO - Using basic auth.
[2023-02-10, 09:35:44 UTC] {databricks.py:97} INFO - example_task in run state: {'life_cycle_state': 'PENDING', 'result_state': '', 'state_message': 'Waiting for cluster'}
[2023-02-10, 09:35:44 UTC] {databricks.py:98} INFO - View run status, Spark UI, and logs at <removed-for-publication>
[2023-02-10, 09:35:44 UTC] {databricks.py:99} INFO - Sleeping for 30 seconds.
[2023-02-10, 09:35:44 UTC] {local_task_job.py:225} WARNING - DagRun timed out after 0:01:05.297413.
[2023-02-10, 09:35:44 UTC] {local_task_job.py:226} WARNING - State of this instance has been externally set to skipped. Terminating instance.
[2023-02-10, 09:35:44 UTC] {process_utils.py:125} INFO - Sending Signals.SIGTERM to group 69373. PIDs of all processes in the group: [69373]
[2023-02-10, 09:35:44 UTC] {process_utils.py:80} INFO - Sending the signal Signals.SIGTERM to group 69373
[2023-02-10, 09:35:44 UTC] {taskinstance.py:1555} ERROR - Received SIGTERM. Terminating subprocesses.
[2023-02-10, 09:35:44 UTC] {base.py:68} INFO - Using connection ID 'databricks_default' for task execution.
[2023-02-10, 09:35:44 UTC] {databricks_base.py:333} INFO - Using basic auth.
[2023-02-10, 09:35:44 UTC] {databricks.py:345} INFO - Task: example_task with run_id: 102832491 was requested to be cancelled.
[2023-02-10, 09:35:44 UTC] {process_utils.py:75} INFO - Process psutil.Process(pid=69373, status='terminated', exitcode=0, started='09:34:41') (69373) terminated with exit code 0
```

Finally, a screenshot of the DAG failure for timing purposes:

![image](https://user-images.githubusercontent.com/3661031/218072255-4ab3f608-8dfc-442a-b59f-c87c4b20f834.png)

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by hussein-awala** [comment]
Duplicate of #27614

Please can check @potiuk explanation in [this comment](https://github.com/apache/airflow/issues/27614#issuecomment-1336132339) to understand why it's currently executed twice (and maybe more in some cases).

**Comment by Oduig** [comment]
Imo it is fair that QoS at-least-once applies here, and that sometimes a scenario can occur where two processes call a callback simultaneously, or a notification delivery succeeded but appeared to the sender as a failure, and it would be delivered again.

However, a callback that is always called twice does not fit at-least-once semantics. In addition, with a BashOperator timeout, the callback is only executed once, so it seems like a solvable problem. Finally, when there is a timeout, the running task is marked skipped rather than failed - only the DAG itself fails, in my eyes this means the task level shouldn't even attempt to call the callback because it itself didn't fail

**Comment by potiuk** [comment]
@Oduig  Does the error occur only for Databricks Submit Run or for other operators as well ?  Can you please double check and report back? 

Also can you please explain in which processes the callbacks are originated - is the same scheduler/dagfile processor that is producing the same stack trace or two different schedulers ? We miss the information on your deployment about those details - we also have no idea how composer internals work, and in such case when there are always duplicate callbacks it seems that it migh be related to some duplication resulting from deployment (but this is a wild guess).. 

It is quite likely that this is maybe even related to some specific deployment issues in 2.3.4 so I also advice you to move to 2.4.3 (which is currently latest in Composer). and report back to us if it hapens there. Also note that in case we will find and release fix here, it will be only released in the latest airflow (2.5.2 in-progress currently), so you will have to upgrade anyway to apply the fix, but more likely this is somehow a deployment issue and you will have to go through regular Composer issue (upgrading to 2.4.3 now to check if the problem persists is likely much faster route for you so I recommend you do it as  first step) 

If you upgrade, reproduce, provide more information on the deployment and origin of those callbacks (if they will happen again after upgrade), then we can likely help you to identify what problem you should report to Composer (as long as we won't find during the process that this is something regular airflow also has).

**Comment by Oduig** [comment]
If I use a `BashOperator` instead, the failure callback is only invoked once. I checked the stack traces by using `traceback.extract_stack()`. It could be related to Composer, we can try to upgrade it. Let me get back to you on this.
